src/engine.rs

```diff
@@ -1126,4 +1135,74 @@ mod tests {
             });
         }
     }
+
+    #[test]
+    fn test_dump_file_or_directory() {
+        let dir = tempfile::Builder::new()
+            .prefix("test_dump_file_or_directory")
+            .tempdir()
+            .unwrap();
+        let entry_data = vec![b'x'; 1024];
+
+        let mut batches = vec![vec![LogBatch::default()]];
+        let mut batch = LogBatch::default();
+        batch
+            .add_entries::<Entry>(7, &generate_entries(1, 11, Some(&entry_data)))
+            .unwrap();
+        batch.add_command(7, Command::Clean);
+        batch.put(7, b"key".to_vec(), b"value".to_vec());
+        batch.delete(7, b"key2".to_vec());
+        batches.push(vec![batch.clone()]);
+        let mut batch2 = LogBatch::default();
+        batch2.put(8, b"key3".to_vec(), b"value".to_vec());
+        batch2
+            .add_entries::<Entry>(8, &generate_entries(5, 15, Some(&entry_data)))
+            .unwrap();
+        batches.push(vec![batch, batch2]);
+
+        let cfg = Config {
+            dir: dir.path().to_str().unwrap().to_owned(),
+            ..Default::default()
+        };
+
+        let engine =
+            RaftLogEngine::open_with_file_builder(cfg, Arc::new(DefaultFileBuilder)).unwrap();
+        for bs in batches.iter_mut() {
+            for batch in bs.iter_mut() {
+                engine.write(batch, false).unwrap();
+            }
+
+            engine.sync().unwrap();
+        }
```

Comment:
Close the engine before you dump it. Or there will be some uninitialized space at the end of file.

Reply:
seems simply use ` drop(engine)` can't fix test error, the test failed again

Reply:
It seems to be a bug. Could you implement `Drop` for the `SinglePipe`? In which we properly close the file by calling `writer.close()`.

Reply:
@tabokie ， OK, let me have a try